### PR TITLE
Account Transaction : Update credential keys

### PR DESCRIPTION
**Pull request by tds-1**
## Purpose
Adds support for Account Transaction ; Update Credential Keys

## Changes
Adds support for Account Transaction ; Update Credential Keys

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

## CLA acceptance

By submitting the contribution I accept the terms and conditions of the
Contributor License Agreement v1.0
- link: https://developers.concordium.com/CLAs/Contributor-License-Agreement-v1.0.pdf

- [x] I accept the above linked CLA.

### CODE REVIEW

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/accountinfo/credential/Key.java [hunk 1]
```diff
@@ -1,20 +1,55 @@
...
     }
+
+    byte[] getSchemeIdBytes() {
+        val buffer = ByteBuffer.allocate(PublicKey.BYTES);
+        switch (this.schemeId.toLowerCase()) {
+            case ED25519:
+                buffer.put(PublicKey.ED25519.getValue());
+                break;
+            default:
+                throw new IllegalStateException("Invalid AccountRequest Type " + this.schemeId.toLowerCase());
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/responses/accountinfo/credential/Key.java:36]: This message is not really correct is it?

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/accountinfo/credential/Key.java [hunk 2]
```diff
@@ -1,20 +1,55 @@
...
+
+import java.nio.ByteBuffer;
 
 @Getter
 @ToString
 public final class Key {
     private final String verifyKey;
     private final String schemeId;
 
+    private final String ED25519 = "ed25519";
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/responses/accountinfo/credential/Key.java:20]: Maybe `ED25519PublicKey` could be used instead?

I don't see any benefits from having `this`, `PublicKey` below and `ED25519PublicKey` representing the same type.

**Review by EmilLa1 — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -951,6 +951,44 @@ try{
 }
 ```
 
+#### Update Credential Keys transaction
```
[the remark this answers is not in the document]
Comment (tds-1) [reply README.md:957]: Done

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/PublicKey.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package com.concordium.sdk.transactions;
+
+import lombok.Getter;
+
+public enum PublicKey {
```
[the remark this answers is not in the document]
Comment (tds-1) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/PublicKey.java:9]: Added

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/accountinfo/credential/Key.java (hunk 1 above)
Comment (tds-1) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/accountinfo/credential/Key.java]: Fixed

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/accountinfo/credential/Key.java (hunk 2 above)
Comment (tds-1) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/accountinfo/credential/Key.java]: Changed, the flow to use ED25519PublicKey

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/CredentialPublicKeys.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import lombok.Getter;
+import lombok.SneakyThrows;
+import lombok.ToString;
+import lombok.val;
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+@Getter
+@ToString
+public class CredentialPublicKeys {
```
[the remark this answers is not in the document]
Comment (tds-1) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/CredentialPublicKeys.java:16]: Added

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateCredentialKeys.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+     * Id of the credential whose keys are to be updated.
+     */
+    private final CredentialRegistrationId credentialRegistrationID;
+
+    /**
+     *  The new public keys
+     */
+    private final CredentialPublicKeys keys;
+
+    private final UInt64 maxEnergyCost;
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateCredentialKeys.java:29]: This should not be necessary. It is possible to pre compute the cost of this type of transaction beforehand. 
See https://github.com/Concordium/concordium-base/blob/a4c5fa5349ee9c5b775dfea55d7e5f8e322be347/rust-src/concordium_base/src/transactions.rs#L1762

**Review by EmilLa1 — CHANGES_REQUESTED** [review]
Looks good. However we don't need to have a `maxEnergyCost` as it can be pre computed based on the keys that are removed and the ones that are added.

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateCredentialKeys.java (hunk 6 above)
Comment (tds-1) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/UpdateCredentialKeys.java]: Changed

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/CredentialPublicKeys.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    private final int threshold;
+
+    CredentialPublicKeys(Map<Index, ED25519PublicKey> keys,
+                         int threshold) {
+        this.keys = keys;
+        this.threshold = threshold;
+    }
+
+    public static CredentialPublicKeys from(Map<Index, ED25519PublicKey> keys, int threshold) {
```
[the remark this answers is not in the document]
Comment (tds-1) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/CredentialPublicKeys.java:40]: Changed

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/PublicKey.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package com.concordium.sdk.transactions;
+
+import lombok.Getter;
+
+/**
+ * Public verification key for a given signature scheme.
+ */
+
+public enum PublicKey {
```
[the remark this answers is not in the document]
Comment (tds-1) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/PublicKey.java]: Changed

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/CredentialPublicKeys.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (EmilLa1) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/CredentialPublicKeys.java:40]: One should not be able to create a `CredentialPublicKeys` with a threshold = 0. It must be greater than zero.

**Review by EmilLa1 — CHANGES_REQUESTED** [review]
Looks good. One comment left that we need to resolve.

**Review by EmilLa1 — APPROVED** [review]
Looks great! GJ

**Pull request merged**